reconcile/utils/instrumented_wrappers.py

```diff
@@ -0,0 +1,81 @@
+from prometheus_client import Counter, Gauge
+
+from sretoolbox.container import Image
+
+
+class InstrumentedImage(Image):
+    """Normal Image that exposes the count of reachouts to external
+    registries.
+
+    It helps us understand the performance of our caches and predict
+    our mirroring-related costs.
+
+    """
+    _registry_reachouts = Counter(
+        name='qontract_reconcile_registry_reachouts',
+        documentation='Number GET requests on public image registries',
+    )
+    #    labelnames=['integrations', 'shards', 'shard_id'])
+
+    def _request_get(self, url):
```

Comment:
We found out that not all requests count agains the quota. requests to `https://auth.docker.io/token` and `https://registry-1.docker.io/v2/<repo>/<image>/tags/list` won't, but requests to `https://registry-1.docker.io/v2/<repo>/<image>/manifests/<tag>` will.

We might then want to either split the caller function per endpoint or track the different endpoints with different metric labels. Main point being to understand how many calls that count against quota we are doing.

Reply:
Which requests do? Just manifest ones? In that case I could override just `_get_manifest`. But if there are any other calls that count against our quota, I'd rather override this one and accept a non-zero baseline. It keeps the code simple and avoids any bugs from delivering a misleading metric.

Reply:
`_get_manifest` calls `_request_get`, which then calls both `/token` and `/manifests`. But only the call against `/manifests` will count against the quota, aparently.

Reply:
>  I'd rather override this one and accept a non-zero baseline. 

I believe it depends on what you want. Do you want to track the number of total requests to the docker API or the number of requests that count against 5k/day quota?

Reply:
In that case, I'll override `_get_manifest` - it'll give the correct numbers. Give me a jiffy.

Reply:
I don't think it will. Now you left out /tags and the related /token calls. For all the API calls, code was good before.

Reply:
> Do you want to track the total number of requests to the docker API or the number of requests that count against 5k/day quota?

It depends. I prefer to count calls that don't add to the quota than leave out calls that do. If we're confident that out of the methods in `Image` only `_get_manifest` adds to the quota, my current version is the right one. Otherwise, indeed, I'll go back to override `_request_get` - and get a number that isn't perfect but is strongly correlated with our quota consumption.

Reply:
`_get_manifest()` calls `/token` and `/manifest`. `/token` doesn't count, `/manifest` does count.

Reply:
What'I'm trying to say is: the way `_request_get` is implemented, it calls `/token` whenever it needs to, and that call doesn't count for the quota.

Reply:
Actually, I was checking the current code. we should be fine now. Sorry about the noise, I will review the latest version when you drop the WiP from it.